File: android/app/src/main/kotlin/com/perfin/app/LeapBridgePlugin.kt
package com.perfin.app

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File

// ─────────────────────────────────────────────────────────────────────────────
// LeapBridgePlugin — Flutter ↔ Liquid AI LEAP SDK bridge (Android)
//
// SETUP STEPS (before the on-device feature works):
//   1. Add the LEAP Android SDK to android/app/build.gradle:
//        dependencies {
//            implementation 'ai.liquid:leap-android:x.y.z'
//        }
//   2. Add the LEAP Maven repository to android/build.gradle if required.
//   3. Uncomment the `import ai.liquid.leap.*` line below.
//   4. Uncomment the LEAP-specific blocks marked TODO throughout this file.
//   5. Obtain a LEAP API key from https://leap.liquid.ai and store it securely
//        (e.g. BuildConfig or Android Keystore) then pass it when initialising
//        the ModelRunner.
// ─────────────────────────────────────────────────────────────────────────────

// import ai.liquid.leap.*  // TODO: uncomment after adding LEAP SDK dependency

class LeapBridgePlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var context: Context

    // TODO: replace Any with ai.liquid.leap.ModelRunner once SDK is added.
    private var modelRunner: Any? = null
    private var eventSink: EventChannel.EventSink? = null

    private val scope = CoroutineScope(Dispatchers.IO)

    // ── FlutterPlugin ─────────────────────────────────────────────────────────

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        methodChannel = MethodChannel(binding.binaryMessenger, "com.perfin.app/leap")
        methodChannel.setMethodCallHandler(this)
        eventChannel = EventChannel(binding.binaryMessenger, "com.perfin.app/leap_stream")
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }

    // ── MethodCallHandler ─────────────────────────────────────────────────────

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "loadModel" -> {
                val path = call.argument<String>("path")
                    ?: return result.error("INVALID_ARGS", "path required", null)
                loadModel(path, result)
            }
            "unloadModel" -> {
                modelRunner = null
                result.success(null)
            }
            "complete" -> {
                val prompt = call.argument<String>("prompt")
                    ?: return result.error("INVALID_ARGS", "prompt required", null)
                val maxTokens = call.argument<Int>("maxTokens") ?: 512
                val temperature = call.argument<Double>("temperature") ?: 0.7
                complete(prompt, maxTokens, temperature, result)
            }
            "completeStream" -> {
                val prompt = call.argument<String>("prompt")
                    ?: return result.error("INVALID_ARGS", "prompt required", null)
                val maxTokens = call.argument<Int>("maxTokens") ?: 512
                val temperature = call.argument<Double>("temperature") ?: 0.7
                completeStream(prompt, maxTokens, temperature)
                result.success(null)
            }
            "getModelInfo" -> {
                if (modelRunner == null) {
                    result.success(mapOf("loaded" to false))
                } else {
                    // TODO: return real metadata from LEAP SDK.
                    result.success(mapOf("loaded" to true, "name" to "LFM2", "contextLength" to 32768))
                }
            }
            else -> result.notImplemented()
        }
    }

    // ── Model loading ─────────────────────────────────────────────────────────

    private fun loadModel(path: String, result: Result) {
        scope.launch {
            // TODO: replace stub with real LEAP SDK call:
            //
            // try {
            //     modelRunner = ModelRunner.fromPath(context, path)
            //     withContext(Dispatchers.Main) { result.success(mapOf("success" to true)) }
            // } catch (e: Exception) {
            //     withContext(Dispatchers.Main) {
            //         result.success(mapOf("success" to false, "error" to e.message))
            //     }
            // }
            //
            // STUB — succeeds if the file exists on disk:
            val fileExists = File(path).exists()
            withContext(Dispatchers.Main) {
                if (fileExists) {
                    modelRunner = path   // placeholder
                    result.success(mapOf("success" to true))
                } else {
                    result.success(mapOf("success" to false, "error" to "Model file not found at: $path"))
                }
            }
        }
    }

    // ── Inference ─────────────────────────────────────────────────────────────

    private fun complete(prompt: String, maxTokens: Int, temperature: Double, result: Result) {
        if (modelRunner == null) {
            result.success("Model is not loaded.")
            return
        }
        scope.launch {
            // TODO: replace stub with real LEAP SDK call:
            //
            // val runner = modelRunner as ModelRunner
            // val conversation = Conversation()
            // conversation.addMessage(Role.User, prompt)
            // val response = runner.generate(conversation, maxTokens = maxTokens)
            // withContext(Dispatchers.Main) { result.success(response.text) }
            //
            // STUB response for development:
            val stub = "[On-device stub] Received: \"${prompt.take(60)}…\""
            withContext(Dispatchers.Main) { result.success(stub) }
        }
    }

    private fun completeStream(prompt: String, maxTokens: Int, temperature: Double) {
        if (modelRunner == null) {
            eventSink?.success("[On-device] Model not loaded.")
            eventSink?.endOfStream()
            return
        }
        scope.launch {
            // TODO: replace stub with real LEAP streaming:
            //
            // val runner = modelRunner as ModelRunner
            // runner.generateStream(prompt, maxTokens).collect { token ->
            //     withContext(Dispatchers.Main) { eventSink?.success(token) }
            // }
            // withContext(Dispatchers.Main) { eventSink?.endOfStream() }
            //
            // STUB — emits prompt words one-by-one:
            val words = "[On-device stub] ${prompt.take(60)}".split(" ")
            for (word in words) {
                Thread.sleep(50)
                withContext(Dispatchers.Main) { eventSink?.success("$word ") }
            }
            withContext(Dispatchers.Main) { eventSink?.endOfStream() }
        }
    }

    // ── EventChannel.StreamHandler ────────────────────────────────────────────

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
}

File: ios/Runner/LeapBridgePlugin.swift
import Flutter
import Foundation

// ─────────────────────────────────────────────────────────────────────────────
// LeapBridgePlugin — Flutter ↔ Liquid AI LEAP SDK bridge (iOS)
//
// SETUP STEPS (before the on-device feature works):
//   1. Add the LEAP iOS SDK via Swift Package Manager in Xcode:
//        File → Add Package Dependencies
//        URL: https://github.com/liquid-ai/leap-swift (replace with real URL)
//   2. Uncomment the `import LEAP` line below.
//   3. Uncomment the LEAP-specific blocks marked TODO throughout this file.
//   4. Obtain a LEAP API key from https://leap.liquid.ai and add it to Info.plist
//        or pass it as an initialisation argument.
// ─────────────────────────────────────────────────────────────────────────────

// import LEAP  // TODO: uncomment after adding the Swift Package

class LeapBridgePlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    // MARK: - Registration

    static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(
            name: "com.perfin.app/leap",
            binaryMessenger: registrar.messenger()
        )
        let eventChannel = FlutterEventChannel(
            name: "com.perfin.app/leap_stream",
            binaryMessenger: registrar.messenger()
        )

        let instance = LeapBridgePlugin()
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
    }

    // MARK: - State

    // TODO: replace Any with the real LEAP ModelRunner type once SDK is added.
    private var modelRunner: Any? = nil
    private var eventSink: FlutterEventSink? = nil

    // MARK: - MethodChannel handler

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {

        case "loadModel":
            guard let args = call.arguments as? [String: Any],
                  let path = args["path"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "path required", details: nil))
                return
            }
            loadModel(path: path, result: result)

        case "unloadModel":
            modelRunner = nil
            result(nil)

        case "complete":
            guard let args = call.arguments as? [String: Any],
                  let prompt = args["prompt"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "prompt required", details: nil))
                return
            }
            let maxTokens = args["maxTokens"] as? Int ?? 512
            let temperature = args["temperature"] as? Double ?? 0.7
            complete(prompt: prompt, maxTokens: maxTokens, temperature: temperature, result: result)

        case "completeStream":
            guard let args = call.arguments as? [String: Any],
                  let prompt = args["prompt"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "prompt required", details: nil))
                return
            }
            let maxTokens = args["maxTokens"] as? Int ?? 512
            let temperature = args["temperature"] as? Double ?? 0.7
            completeStream(prompt: prompt, maxTokens: maxTokens, temperature: temperature)
            result(nil)

        case "getModelInfo":
            guard modelRunner != nil else {
                result(["loaded": false])
                return
            }
            // TODO: return real model metadata from LEAP SDK.
            result(["loaded": true, "name": "LFM2", "contextLength": 32768])

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Model loading

    private func loadModel(path: String, result: @escaping FlutterResult) {
        DispatchQueue.global(qos: .userInitiated).async {
            // TODO: replace stub with real LEAP SDK call:
            //
            // do {
            //     self.modelRunner = try ModelRunner(modelPath: path)
            //     DispatchQueue.main.async { result(["success": true]) }
            // } catch {
            //     DispatchQueue.main.async {
            //         result(["success": false, "error": error.localizedDescription])
            //     }
            // }
            //
            // STUB — succeeds without a real model so Dart UI can be tested:
            let fileExists = FileManager.default.fileExists(atPath: path)
            DispatchQueue.main.async {
                if fileExists {
                    self.modelRunner = path  // placeholder
                    result(["success": true])
                } else {
                    result(["success": false, "error": "Model file not found at: \(path)"])
                }
            }
        }
    }

    // MARK: - Inference

    private func complete(
        prompt: String,
        maxTokens: Int,
        temperature: Double,
        result: @escaping FlutterResult
    ) {
        guard modelRunner != nil else {
            result("Model is not loaded.")
            return
        }
        DispatchQueue.global(qos: .userInitiated).async {
            // TODO: replace stub with real LEAP SDK call:
            //
            // let runner = self.modelRunner as! ModelRunner
            // let conversation = Conversation()
            // conversation.addMessage(role: .user, content: prompt)
            // let response = try await runner.generate(conversation, maxTokens: maxTokens)
            // DispatchQueue.main.async { result(response.text) }
            //
            // STUB response for development:
            let stub = "[On-device stub] Received: \"\(prompt.prefix(60))…\""
            DispatchQueue.main.async { result(stub) }
        }
    }

    private func completeStream(prompt: String, maxTokens: Int, temperature: Double) {
        guard modelRunner != nil else {
            eventSink?("[On-device] Model not loaded.")
            eventSink?(FlutterEndOfEventStream)
            return
        }
        DispatchQueue.global(qos: .userInitiated).async {
            // TODO: replace stub with real LEAP streaming:
            //
            // let runner = self.modelRunner as! ModelRunner
            // for try await token in runner.generateStream(prompt, maxTokens: maxTokens) {
            //     DispatchQueue.main.async { self.eventSink?(token) }
            // }
            // DispatchQueue.main.async { self.eventSink?(FlutterEndOfEventStream) }
            //
            // STUB — emits the prompt back word-by-word:
            let words = "[On-device stub] \(prompt.prefix(60))".split(separator: " ")
            for word in words {
                Thread.sleep(forTimeInterval: 0.05)
                DispatchQueue.main.async { self.eventSink?(String(word) + " ") }
            }
            DispatchQueue.main.async { self.eventSink?(FlutterEndOfEventStream) }
        }
    }

    // MARK: - FlutterStreamHandler

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}
